snowpack/src/commands/dev.ts

```diff
@@ -120,6 +121,44 @@ class NotFoundError extends Error {
   }
 }
 
+const noOpHeadersTransformer: HeadersTransformer = (_req, proposed) => proposed;
+
+/**
+ * Decorates res.writeHead, such that a user-supplied HeadersTransformer may inspect
+ * the proposed response headers, and (with reference to the request): optionally propose
+ * alternative headers.
+ * This enables users to apply their own CORS or security policies, and to override decisions
+ * about Content-Type.
+ */
+function decorateWriteHead(
+  req: http.IncomingMessage,
+  res: http.ServerResponse,
+  headersTransformer: HeadersTransformer
+): void {
```

Comment:
agreed that it'd be good to move the side-effect to happen at call-site for more visibility.

though, I'm thinking that the "monkey-patch `writeHead()`" idea only solves a small problem (i.e. it ensures downstream code doesn't need to know about header transformation, and can just writeHead like it always did). I think the bigger problem it creates is that `HeadersTransformer` isn't a very nice interface for the user (since they need to be prepared for dictionary/array/null headers, and are allowed to return in any of those shapes).

I think better design would be:

- `dev.ts` should use a helper function instead of direct `res.writeHead()`
  - this means about 7 callsites need updating (same way as in https://github.com/snowpackjs/snowpack/pull/3030)
- HeadersTransformer should use dictionary-style input and output only.

so, basically going back to the implementation in the previous PR, except configurability would continue to be exposed via `RouteConfigObject`, rather than `SnowpackPlugin['devServer']`.